libservo: Add initial `NetworkManager::cache_entries` API

janvarga opened the pull request:
Add initial support in `NetworkManager` for listing entries currently stored
in the HTTP cache. Each returned entry is identified by its cache key (URL).
    
The necessary support has been added to the net crate to expose cache entry
information.

Testing: A new integration test has been added.

janvarga committed:
libservo: Add support to `NetworkManager` for listing HTTP cache entries

Add support in `NetworkManager` for listing entries currently stored in the
HTTP cache. Each returned entry is identified by its cache key (URL).

The necessary support has been added to the net crate to expose cache entry
information.

Signed-off-by: Jan Varga <jvarga@igalia.com>

mrobinson reviewed (commented):
Looks like this is still in progress, but here are a few drive by comments:

mrobinson commented on components/servo/network_manager.rs:
```diff
@@ -2,8 +2,25 @@
+impl CacheEntry {
+    pub fn new(key: String) -> CacheEntry {
+        CacheEntry { key }
+    }
```
```suggestion
    pub fn new(key: String) -> Self {
        Self { key }
    }
```

janvarga replied:
Ok, I now see that's the preferred style and makes potential renaming easier.

mrobinson commented on components/shared/net/lib.rs:
```diff
@@ -646,6 +656,17 @@ impl SiteDescriptor {
+
+impl CacheEntryDescriptor {
+    pub fn new(key: String) -> Self {
+        CacheEntryDescriptor { key }
```
```suggestion
        Self { key }
```

mrobinson commented on components/servo/tests/network_manager.rs:
```diff
@@ -8,12 +8,57 @@ use std::rc::Rc;
+    let cache_entries = network_manager.cache_entries();
+    assert_eq!(
+        &cache_entries,
+        &[CacheEntry::new(format!("http://localhost:{}/", port)),]
```
```suggestion
        &[CacheEntry::new(format!("http://localhost:{port}/")),]
```

janvarga replied:
Nice

mrobinson reviewed (approved):
Nice I have a few nits below:

janvarga commented on components/servo/network_manager.rs:
```diff
@@ -25,6 +42,32 @@ impl NetworkManager {
+            entries.insert(private_entry.key);
+        }
+
+        entries.into_iter().map(CacheEntry::new).collect()
```
Ok, I addressed all other suggestions locally, but this one seems to be tricky. I'll try if it compiles with some modifications.

janvarga replied:
So it seems the union would requires some cloning, but I discovered that `chain` can be used to vectors/iterators...
The code is a bit more verbose, but it at least avoids the mutability and also some allocations/re-allocations.

janvarga committed:
libservo: Address review feedback for `NetworkManager::cache_entries` API

This patch incorporates review feedback by applying the following cleanups
and refinements:

- Rename `CoreResourceMsg::ListCacheEntries` to `GetCacheEntries`
- Use `Self` in `new` functions
- Use more efficient de-duplication in `NetworkManager::cache_entries`
- Use a more compact form in `format!`

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga commented on components/servo/network_manager.rs:
```diff
@@ -2,8 +2,25 @@
+
+    pub fn key(&self) -> String {
+        self.key.clone()
+    }
```
Ok

mrobinson reviewed (approved):
Thanks!

mrobinson committed:
Update components/servo/network_manager.rs

Signed-off-by: Martin Robinson <mrobinson@igalia.com>
